idaes/apps/caprese/examples/cstr_mhe.py

```diff
@@ -0,0 +1,229 @@
+##############################################################################
+# Institute for the Design of Advanced Energy Systems Process Systems
+# Engineering Framework (IDAES PSE Framework) Copyright (c) 2018-2019, by the
+# software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia
+# University Research Corporation, et al. All rights reserved.
+#
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+"""
+Example for Caprese's module for MHE.
+"""
+import random
+from idaes.apps.caprese.dynamic_builder import DynamicSim
+from idaes.apps.caprese.util import apply_noise_with_bounds
+from pyomo.environ import SolverFactory, Reference
+from pyomo.dae.initialization import solve_consistent_initial_conditions
+# import idaes.logger as idaeslog
+from idaes.apps.caprese.examples.cstr_model import make_model
+from idaes.apps.caprese.data_manager import EstimatorDataManager
+
+__author__ = "Kuan-Han Lin"
+
+
+
+# See if ipopt is available and set up solver
+if SolverFactory('ipopt').available():
+    solver = SolverFactory('ipopt')
+    solver.options = {
+            'tol': 1e-6,
+            'bound_push': 1e-8,
+            'halt_on_ampl_error': 'yes',
+            'linear_solver': 'ma57',
+            }
+else:
+    solver = None
+
+
+def main():
+    m_estimator = make_model(horizon=8, ntfe=16, ntcp=2, bounds=True)
+    sample_time = 0.5
+    m_plant = make_model(horizon=sample_time, ntfe=5, ntcp=2)
+    time_plant = m_plant.fs.time
+
+    simulation_horizon = 20
+    n_samples_to_simulate = round(simulation_horizon/sample_time)
+
+    samples_to_simulate = [time_plant.first() + i*sample_time
+                           for i in range(1, n_samples_to_simulate)]
+
+    # We must identify for the estimator which variables are our
+    # inputs and measurements.
+    inputs = [
+            m_plant.fs.mixer.S_inlet.flow_vol[0],
+            m_plant.fs.mixer.E_inlet.flow_vol[0],
+            ]
+    measurements = [
+            m_plant.fs.cstr.outlet.conc_mol[0, 'C'],
+            m_plant.fs.cstr.outlet.conc_mol[0, 'E'],
+            m_plant.fs.cstr.outlet.conc_mol[0, 'S'],
+            m_plant.fs.cstr.outlet.conc_mol[0, 'P'],
+            m_plant.fs.cstr.outlet.temperature[0],
+            m_plant.fs.cstr.volume[0],
+            ]
+    
+    # Construct the "MHE simulator" object
+    mhe = DynamicSim(
```

Comment:
Something to think about: What if we wanted to do MHE (or NMPC) without a plant model, i.e. just with a stream of data or a black-box (commercial) simulator?